src/servers/http_server.cc

```diff
@@ -321,280 +142,291 @@ class HTTPAPIServer : public HTTPServerImpl {
   std::string mode;
   if ((health_uri.empty()) ||
       (!RE2::FullMatch(health_uri, health_regex_, &mode))) {
-    evhtp_send_reply(req, EVHTP_RES_BADREQ);
-    return;
+    req->res.status = 400;
+    req->res.reason = "Bad request";
+    h2o_add_header(
+        &req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, NULL,
+        H2O_STRLIT("text/plain"));
+    h2o_generator_t generator = {NULL, NULL};
+    h2o_iovec_t body = h2o_strdup(&req->pool, "", SIZE_MAX);
+    h2o_start_response(req, &generator);
+    h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
+    return 0;
   }
 
   TRTSERVER_Error* err = nullptr;
   bool health = false;
 
   if (mode == "live") {
-    err = TRTSERVER_ServerIsLive(server_.get(), &health);
+    err = TRTSERVER_ServerIsLive(self->http_server->server_.get(), &health);
   } else if (mode == "ready") {
-    err = TRTSERVER_ServerIsReady(server_.get(), &health);
-  } else {
-    err = TRTSERVER_ErrorNew(
-        TRTSERVER_ERROR_UNKNOWN,
-        std::string("unknown health mode '" + mode + "'").c_str());
+    err = TRTSERVER_ServerIsReady(self->http_server->server_.get(), &health);
   }
 
+  h2o_generator_t generator = {NULL, NULL};
+  h2o_iovec_t body = h2o_strdup(&req->pool, "", SIZE_MAX);
+
   RequestStatus request_status;
   RequestStatusUtil::Create(
       &request_status, err, RequestStatusUtil::NextUniqueRequestId(),
-      server_id_);
-
-  evhtp_headers_add_header(
-      req->headers_out,
-      evhtp_header_new(
-          kStatusHTTPHeader, request_status.ShortDebugString().c_str(), 1, 1));
-
-  evhtp_send_reply(
-      req, (health && (err == nullptr)) ? EVHTP_RES_OK : EVHTP_RES_BADREQ);
+      self->http_server->server_id_);
+
+  // Add NV-Status header
+  std::string status_header = std::string(kStatusHTTPHeader);
+  h2o_iovec_t status_header_content = h2o_strdup(
+      &req->pool, request_status.ShortDebugString().c_str(), SIZE_MAX);
+  h2o_add_header_by_str(
+      &req->pool, &req->res.headers, status_header.c_str(),
+      status_header.size(), 0, NULL, status_header_content.base,
+      status_header_content.len);
+  h2o_add_header(
+      &req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, NULL,
+      H2O_STRLIT("text/plain"));
+
+  if (health && (err == nullptr)) {
+    req->res.status = 200;
+    req->res.reason = "OK";
+    req->res.content_length = body.len;
+  } else {
+    req->res.status = 400;
+    req->res.reason = "Bad request";
+  }
 
-  TRTSERVER_ErrorDelete(err);
+  h2o_start_response(req, &generator);
+  h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
+  return 0;
 }
 
-void
-HTTPAPIServer::HandleStatus(evhtp_request_t* req, const std::string& status_uri)
+int
+HTTPAPIServer::Status(h2o_handler_t* handler_self, h2o_req_t* req)
 {
-  if (req->method != htp_method_GET) {
-    evhtp_send_reply(req, EVHTP_RES_METHNALLOWED);
-    return;
+  h2o_custom_req_handler_t* self = (h2o_custom_req_handler_t*)handler_self;
+  re2::RE2 status_regex(R"(/api/status(/(.*))?)");
```

Comment:
I would think there is a way to register a regex and have h2o do the match for the correct handler and then deliver the matched groups to the handler